tests/main/views/test_audits.py

```diff
@@ -50,8 +54,204 @@ def add_audit_events_with_db_object(self):
+        # pseudo-ids from 0-4 inclusive
+        chain.from_iterable(
+            ((serv_aeps, supp_aeps, req_params, expected_resp_events) for req_params, expected_resp_events in req_cases)
+            for serv_aeps, supp_aeps, req_cases in (
+                (
+                    (  # service_audit_event_params
+                        (0, AuditTypes.update_service, datetime(2010, 6, 6), None,),
+                        (0, AuditTypes.update_service, datetime(2010, 6, 7), None,),
+                        (4, AuditTypes.update_service, datetime(2010, 6, 2), None,),
+                    ),
+                    (  # supplier_audit_event_params
+                        (0, AuditTypes.supplier_update, datetime(2010, 6, 6), None,),
+                    ),
+                    (  # and now a series of req_cases - pairs of (req_params, expected_resp_events) to test against
+                       # the above db scenario. these get flattened out into concrete test scenarios by
+                       # chain.from_iterable above before they reach pytest's parametrization
+                        (
+                            {"earliest_for_each_object": "true", "latest_first": "true"},
+                            (3, 0, 2,),
+                        ),
+                        (
+                            {"earliest_for_each_object": "true", "latest_first": "false"},
+                            (2, 0, 3,),
+                        ),
+                        (
+                            {"earliest_for_each_object": "true", "audit-type": "supplier_update"},
+                            (3,),
+                        ),
+                    ),
+                ),
+                (
+                    (  # service_audit_event_params
+                        (0, AuditTypes.update_service, datetime(2011, 6, 6), None,),
+                        (0, AuditTypes.update_service_status, datetime(2011, 8, 2), None,),
+                        (1, AuditTypes.update_service, datetime(2011, 8, 6, 12), datetime(2011, 9, 2),),
+                        (1, AuditTypes.update_service, datetime(2011, 8, 6, 9), datetime(2011, 9, 1),),
+                        (0, AuditTypes.update_service, datetime(2010, 8, 4), datetime(2011, 9, 2),),
+                        (3, AuditTypes.update_service, datetime(2014, 6, 6), None,),
+                    ),
+                    (  # supplier_audit_event_params
+                        (1, AuditTypes.supplier_update, datetime(2010, 2, 6), None,),
+                        (0, AuditTypes.supplier_update, datetime(2010, 6, 6), None,),
+                        (1, AuditTypes.supplier_update, datetime(2010, 2, 6), None,),
+                    ),
+                    (  # series of req_cases for the above db scenario
+                        (
+                            {"earliest_for_each_object": "true"},
+                            (6, 7, 4, 3, 5,),
```

Comment:
Again, does the appearance of `6` rather than `8` for supplier_updates to supplier #1 just depend on order of insertion?